### PR TITLE
Add some unit tests

### PR DESCRIPTION
**Pull request by davschneller**
* fix a bug where `importlib` instead of `importlib.util` is imported
* fix parsing empty sparsity patterns during input
* allow `F32` and `F64` as precision strings (to support the changes in SeisSol upstream)
* add some unit tests. Disclaimer: those are AI-generated using Opus 4.7, and modified by me afterwards. (I also generated them for #100 in conjunction; so if there are any remnants here w.r.t. that PR—let me know) Thus, they now also contains comments and explanations of what's happening there.

### CODE REVIEW

File: tests/pytest/test_ast_node.py [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+            m=Indices("i", (3,)),
+            n=Indices("k", (5,)),
+            k=Indices("j", (4,)),
+        )
+        assert log.transA() is True
+        assert log.transB() is False
+
+
+# ---------------------------------------------------------------------------
+# FusedGEMMs - list-like container of LoGs
```
Comment (vikaskurapati) [inline tests/pytest/test_ast_node.py]: What is a LoG here? Is it a list of GEMMs?

File: tests/pytest/test_controlflow.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+)
+from yateto.controlflow.visitor import AST2ControlFlow
+
+
+# ---------------------------------------------------------------------------
+# Helpers
+# ---------------------------------------------------------------------------
+
+
+def _lower_to_cfg(kernel, arch):
```
Comment (vikaskurapati) [inline tests/pytest/test_controlflow.py:70]: I was under the impression we use `_` for private variables and methods within classes to highlight they are supposed to be private. Do we also use them for helper functions? And does the `_` here mention that it is not supposed to be used elsewhere?

File: tests/pytest/test_cost.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        B = Tensor("B", (N, N), spp=diag)
+        C = Tensor("C", (N, N))
+        kernel = C["ij"] <= A["ik"] * B["kj"]
+        kernel = _lower_to_product_tree(kernel, estimator_cls=ExactCost)
+        cost = ExactCost().estimate(kernel)
+        # A dense 4x4 matmul costs 112 in our accounting (2*N^3 - N^2 = 128-16=112).
+        # Diagonal-of-diagonal must be much cheaper - exactly N products
+        # (the contraction collapses to elementwise pairing).
+        # We expect something a lot less than 112 - let's require <= 16.
+        assert cost <= 16
```
Comment (vikaskurapati) [inline tests/pytest/test_cost.py]: If I understand this correctly, 16 here is arbitrary. Am I correct? If that is the case, can we guarantee that the cost is always less than or equal to 16? I.e., are there possible edge cases where this is not satisfied?

File: tests/pytest/test_generator.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+# ---------------------------------------------------------------------------
+
+
+class TestKernelPrefetch:
+    def test_accepts_tensor(self):
+        A = Tensor("A", (4, 4))
+        P = Tensor("P", (4, 4))
+        kernel = Kernel("k", A["ij"] <= A["ij"], prefetch=P)
+        # _prefetch is stored as a list internally.
+        assert kernel._prefetch == [P]
```
Comment (vikaskurapati) [inline tests/pytest/test_generator.py]: Are we not supposed to treat `_prefetch` like a private variable and not access it directly from outside? Is it worth writing a getter function for this?

File: tests/pytest/test_ast_node.py (hunk 1 above)
Comment (davschneller) [reply tests/pytest/test_ast_node.py]: LoG == Loop over GEMM. I.e. you execute multiple GEMM operations in series to model more general tensor contractions of two tensors.

The FusedGEMM operation in essence lists multiple matrix multiplications/GEMMs to become one GPU kernel for ChainForge or tinytc (not TensorForge, that one works differently). ... And yes, it's more strictly matrix multiplications right now (without any other more general tensor contraction-ish things); so probably that comment might need an extra hint.

File: tests/pytest/test_generator.py (hunk 4 above)
Comment (davschneller) [reply tests/pytest/test_generator.py]: Yeah, a getter sounds good I think

File: tests/pytest/test_controlflow.py (hunk 2 above)
Comment (davschneller) [reply tests/pytest/test_controlflow.py:70]: Yes; it's supposed to be private to the file. Like anonymous namespaces in C++ (or static functions outside of classes).

**Review by vikaskurapati — APPROVED** [review]
LGTM. Could be merged after addressing the comments.

File: tests/pytest/test_cost.py (hunk 3 above)
Comment (davschneller) [reply tests/pytest/test_cost.py]: The test basically multiplies a 4×4 diagonal matrix with another 4×4 diagonal matrix—i.e. we get 4 multiplications (or FMAs against 0—which get subtracted away in the calculation).
I checked it—we get exactly 4 as a result. I've updated the test to require 4.

File: yateto/arch.py [hunk 5]
```diff
@@ -69,11 +69,11 @@ def __init__(self,
...
 
     self.precision = precision.upper()
-    if self.precision == 'D':
+    if self.precision in ('D', 'F64'):
       self.bytesPerReal = 8
       self.typename = 'double'
       self.epsilon = 2.22e-16
-    elif self.precision == 'S':
+    elif self.precision in ('S', 'F32'):
       self.bytesPerReal = 4
```
[the remark this answers is not in the document]
Comment (davschneller) [reply yateto/arch.py:79]: Done

File: tests/pytest/test_type.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+        assert arr[0, 0] == pytest.approx(1.5)
+        assert arr[1, 2] == pytest.approx(-2.0)
+        # All other entries remain zero.
+        arr[0, 0] = 0
+        arr[1, 2] = 0
+        assert np.all(arr == 0)
+
+    def test_bad_sparsity_shape_raises(self):
+        wrong = np.ones((2, 2), dtype=bool)
+        with pytest.raises(Exception):
```
[the remark this answers is not in the document]
Comment (davschneller) [reply tests/pytest/test_type.py]: Done (the `Exception` hid another error in the construction of the `ValueError`)

File: .github/workflows/yateto-pytest.yml [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+                 --cov-report=xml \
+                 --cov-report=term-missing \
+                 --cov-branch
+
+    - name: Upload coverage
+      if: matrix.python-version == '3.12'
+      uses: actions/upload-artifact@v4
+      with:
+        name: python-coverage
+        path: tests/python/coverage.xml
```
[the remark this answers is not in the document]
Comment (davschneller) [reply .github/workflows/yateto-pytest.yml]: Done
